### ISSUE TITLE
SFTP issue

### ISSUE THREAD

**Issue by paoloap** [wa2c/cifs-documents-provider#83]
**Describe the bug**
I have an SSH shared folder on my home server and I can access to it through a password-less account (but with an RSA key).
On my smartphone (Pixel 8a with GrapheneOS) I regularly installed and configured the app from F-Droid to be able to browse the folder with the default "Files" application.
The configuration looks OK, in fact if I tap on the "Folder" button from app's configuration I can regularly see the 4 sub-directories into the shared folder.
But into "Files" application, entering into my file server's folder, I see all the folders named like one of them, and I can't navigate into.

**To Reproduce**
1. Install the app
2. Add connection named "Fileserver", type SSHFTP (Apache Commons), insert IP, port, username, leave blank password input, attach the SSH key, select the folder on remote server (on which the user has rwx rights)
3. Save
4. Click on the folder icon in the folder section to verify I can actully see the subfolders with the credentials (OK)
5. Go into AOSP "Files" app
6. Select "CIFS Documents Provider" from left menu
7. Tap on "Fileserver" button
8. Result like in the screenshot appended

**Screenshots**

![93aa3bd560cd4676af1cf89f2bbbf13f](https://github.com/user-attachments/assets/3caf155d-2e41-48c9-be56-cee55cba4f6d)

**CIFS Documents Provider**
 - App Versin: [e.g. 2.0.0]
 - Connection Settings
   - Storage Type: [e.g. SMB2,3(JCIFS-NG)]
   - Authantication: [e.g. Guest, Anonymous, User&Password]
   - Safe Data Taransfer: [e.g. yes]
   - Other: ...
 - App Settings
   - Show notification when opening files: [e.g. no]
   - Use as local storage: [e.g. yes]
   - Other: ...

**App installed device (please complete the following information):**
 - OS: GrapheneOS
 - Device: Pixel 8a
 - Cliant App: CIFS Document Provider 2.3.0 (from F-Droid)
**Server (please complete the following information):**
 - OS: Debian 12 VM Guest
 - Server Software: SSH

**Comment by RebelliousX** [comment]
There is a another issue related to RSA keys that might or might not be related. But to pinpoint the issue if it is related to RSA or not, can you test ed25519 keys instead if possible? I use ed25519 keys with passphrases too and have no issues.

**Comment by paoloap** [comment]
Sorry if I answer only now, I was on vacations and had github logged out!
I just tested after having generated a new ed25519 key pair.
Unfortunately... no luck. Now I want to check if by some case I set wrong files/folders permissions/ownership (but I think everything is ok on that side too...)

**Comment by RebelliousX** [comment]
Sorry for the late reply, I really can't replicate the issue. I am guessing it is a permission issue. 

I am not sure how QNAP NAS system  works. But I use TrueNAS. And despite I set a `Home Directory` for example, `/mnt/MyDataPool/UsersDataset/MyUsername` as a home directory for the user logging in SFTP session, Cifs Document Provider defaults to the root directory `/` instead which displays all directories and system files for the Linux file system. I believe this the correct behavior is to use the home directory of the current user as the starting directory, since all file managers that support SFTP (SSH) do that. And when I run Midnight Commander file manager `mc` command from terminal when doing SSH session, it defaults to the home directory and not the root directory.

 It is possible under QNAP you don't have permission to view system files or read them for the user you are using. Or possibly change the character encoding from default UTF-8 to something else and try again?

**Comment by k9janer** [comment]
Hello, I have the same issue using SFTP but WITHOUT a Private Key.

The issue occurs for me, when I use a connection to a subfolder on my NAS in the Android files app.

So the connection to the root directory (xxx.xxx.xxx.xx/) works fine for me but connecting to (xxx.xxx.xxx.xx/directory jonas/documents and files/documents/) makes the android files app display one single file that is in that directory multiple times and I can't see any of the subfolders. I assume that the file is displayed as often as there are files in the same directory. Just like paoloap I can see the files and subfolders in that directory just fine when creating the connection in CIFS DP.

The files app also acts weird, for example choosing "select all" from the menu will only select one of the many identical files. Could this be a communication error between CIFS DP and the android files app/the files app that is included in GrapheneOS?

Using the connection that leads to the root directory, I can navigate to /directory jonas/documents and files/documents/ and it will display the directory without issues.

System:

CIFS DP is installed on:

Google Pixel Tablet
Latest GrapheneOS (Android 15 QPR2)

The SFTP server runs on my stock Synology NAS DS220j with the latest DSM 7.2.2

**Comment by YanceyChiew** [comment]
I can't reproduce this bug.

> The issue occurs for me, when I use a connection to a subfolder on my NAS in the Android files app.

I used the path related to the user's home directory when configuring the connection, like `'sftp://<hostname>/home/<username>/'`. After referring to your comment, I added the name of the subdirectory at the end, such as '/go', which is also accessible normally.

I think it might have something to do with the server behaving abnormally or permission checks.

I tested `OpenSSH_10.0p2 Debian-2, OpenSSL 3.5.0 8 Apr 2025` and `OpenSSH_9.2p1 Debian-2+deb12u5, OpenSSL 3.0.15 3 Sep 2024` with different paths, both work well.

If your SSH service implementation is different or too old, please try a newer version of the OpenSSH service.

**Comment by k9janer** [comment]
Okay, I just checked on my Synology DS220j running the latest DSM 7.2.2. Unfortunately it looks like it's running a very old version of both OpenSSH and OpenSSL. Since this is a closed system, I unfortunately can't update the components.

OpenSSH_8.2p1, OpenSSL 1.1.1u  30 May 2023

**Comment by RebelliousX** [comment]
Today I encountered this issue when I was recreating SFTP connection in the app. It happens at random. It is going to be hard trying to pinpoint the cause of this. But it is not a permanent issue. I forced closed the app and I tried again with few different things like instead of importing the private key, I referred to the file and vice versa, the issue happened either case anyways. But eventually it fixed itself.

It kept happening few times when I tried to create new connections. I noticed that when I tried to set the base folder for the connection, like usual, it points to the root "/" of my NAS. But after tapping on the `mnt` folder to go to my zfs mounted pool directory, it kept pointing to the `/` directory, no matter how many times I tap on the `mnt` folder.

If I save the connection settings as faulty as it is at this point (using root directory for the base folder) and the `Files` app then displayed all folders with the same randomly selected file / folder from the list repeatedly the same.

Long story short, I don't think it is an issue with the SFTP implementation, I believe there is a subtle corruption that happens to the data store / connection settings somewhere that happens sometimes.

**Comment by Wolf480pl** [comment]
Leaving the port empty (thus the storage URI being like `sftp://nas.lan/srv/archive` instead of `sftp://nas.lan:22/srv/archive`) fixed it for me. I suspect the colon may be breking something.

**Comment by ailinician** [comment]
All good now!> Leaving the port empty (thus the storage URI being like `sftp://nas.lan/srv/archive` instead of `sftp://nas.lan:22/srv/archive`) fixed it for me. I suspect the colon may be breking something.

**Comment by RedKage** [comment]
Confirmed removing the port also fixes the issue for me

**Comment by DexrnZacAttack** [comment]
> Leaving the port empty (thus the storage URI being like `sftp://nas.lan/srv/archive` instead of `sftp://nas.lan:22/srv/archive`) fixed it for me. I suspect the colon may be breking something.

Yep, fixed it for me too.
